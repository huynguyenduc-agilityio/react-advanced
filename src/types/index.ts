export * from './user';
export * from './router';
export * from './pagination';
export * from './queryKey';
export * from './httpsRequest';
export * from './table';
export * from './status';
export * from './type';
