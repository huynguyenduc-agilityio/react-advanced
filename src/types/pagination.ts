export interface PaginationTableType {
  currentPage: number;
  totalPage: number;
}

export interface TPagination<T> {
  meta: {
    pagination: {
      page: number;
      pageSize: number;
      pageCount: number;
      total: number;
    };
  };
  items: T[];
}
