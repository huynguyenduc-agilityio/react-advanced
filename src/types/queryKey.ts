export type TQueryKey = {
  page: number;
  pageSize: number;
  filter?: Record<string, string>;
};
