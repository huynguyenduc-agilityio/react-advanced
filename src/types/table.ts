import { ElementType, JSX } from 'react';

export type TDataSource = {
  id: string;
  [key: string]: string | string[] | number | boolean | undefined;
};

export type THeaderTable = {
  title?: string;
  key?: string;
  icon?: ElementType;
  isSort?: boolean;
  renderBody?: (_: TDataSource, index?: number) => JSX.Element;
};

export type TCellTable = {
  key?: string;
  content?: string | number | boolean | JSX.Element | string[] | undefined;
};
