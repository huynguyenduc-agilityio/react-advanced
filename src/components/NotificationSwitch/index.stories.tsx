import { useState } from 'react';
import { Meta, StoryObj } from '@storybook/react';

// Types
import { NotificationType } from '@/types';

import NotificationSwitch from '.';

const meta: Meta<typeof NotificationSwitch> = {
  title: 'Components/NotificationSwitch',
  component: NotificationSwitch,
  argTypes: {
    value: {
      control: 'radio',
      options: [NotificationType.InApp, NotificationType.Email],
    },
  },
  parameters: {
    layout: 'centered',
  },

  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof NotificationSwitch>;

interface WithStateProps {
  value: NotificationType;
  onChange?: (value: NotificationType) => void;
}

const WithState: React.FC<WithStateProps> = ({ value: initialValue }) => {
  const [value, setValue] = useState<NotificationType>(initialValue);

  return <NotificationSwitch value={value} onChange={setValue} />;
};

export const Default: Story = {
  render: (args) => <WithState value={args.value ?? NotificationType.InApp} />,
  args: { value: NotificationType.InApp },
};

export const EmailSelected: Story = {
  render: (args) => <WithState value={args.value ?? NotificationType.Email} />,
  args: { value: NotificationType.Email },
};
