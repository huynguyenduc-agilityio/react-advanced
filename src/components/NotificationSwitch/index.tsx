import {
  Switch as ChakraSwitch,
  Flex,
  Icon,
  SwitchProps,
  Text,
} from '@chakra-ui/react';
import { forwardRef } from 'react';

// Icons
import { HiMail } from 'react-icons/hi';
import { DevicePhoneIcon } from '@/components';

// Types
import { NotificationType } from '@/types';

type TNotificationSwitchProps = Omit<SwitchProps, 'onChange'> & {
  name?: string;
  value?: NotificationType;
  onChange: (value: NotificationType) => void;
};

const NotificationSwitch = forwardRef<
  HTMLInputElement,
  TNotificationSwitchProps
>(({ name, value = NotificationType.InApp, onChange, ...rest }, ref) => {
  const isChecked = value === NotificationType.Email;

  const handleToggle = () => {
    const newValue = isChecked
      ? NotificationType.InApp
      : NotificationType.Email;

    onChange(newValue);
  };

  return (
    <Flex
      minW="143px"
      align="center"
      justify="center"
      cursor="pointer"
      onClick={handleToggle}
    >
      <Flex
        align="center"
        justifyContent="center"
        gap={2}
        w="full"
        py="6px"
        borderStartRadius="base"
        bg={!isChecked ? 'primary' : 'darkBlue'}
        color={!isChecked ? 'white' : 'pastelBlue'}
        transition="background 0.3s"
      >
        <Icon as={DevicePhoneIcon} />
        <Text fontSize="sm">In-app</Text>
      </Flex>
      <ChakraSwitch
        ref={ref}
        hidden
        isChecked={isChecked}
        onChange={handleToggle}
        name={name}
        {...rest}
      />
      <Flex
        align="center"
        justifyContent="center"
        gap={2}
        w="full"
        py="6px"
        borderEndRadius="base"
        bg={isChecked ? 'primary' : 'darkBlue'}
        color={isChecked ? 'white' : 'pastelBlue'}
        transition="background 0.3s"
      >
        <Icon as={HiMail} />
        <Text fontSize="sm">Email</Text>
      </Flex>
    </Flex>
  );
});

export default NotificationSwitch;
