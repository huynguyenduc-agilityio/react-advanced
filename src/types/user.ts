import { NotificationType } from '@/types';

export interface IUserPersonalInfo {
  name: string;
  avatar?: string;
  email: string;
  description?: string;
  phone: string;
  position: string;
  location: string;
  website: string;
  company: string;
  status?: string;
}

export interface IUserTeamInfo {
  teamName: string;
  rank: string;
  office: string;
  teamMail: string;
}

export interface IUserBillInfo {
  payment: string;
  billName: string;
  billAddress: string;
  state: string;
  zipCode: string;
}

export interface IUserNotifications {
  mentionMessage?: NotificationType;
  replyMessage?: NotificationType;
  assignTask?: NotificationType;
  taskOverdue?: NotificationType;
  dailySummary?: NotificationType;
  weeklySummary?: NotificationType;
  monthlySummary?: NotificationType;
  annuallySummary?: NotificationType;
}

export interface IUserModel
  extends IUserPersonalInfo,
    IUserTeamInfo,
    IUserBillInfo,
    IUserNotifications {
  id?: string;
}

export interface IUsers {
  meta: {
    pagination: {
      page: number;
      pageSize: number;
      pageCount: number;
      total: number;
    };
  };
  items: IUserModel[];
}
