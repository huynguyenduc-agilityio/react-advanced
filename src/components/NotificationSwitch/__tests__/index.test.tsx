import { render, screen, fireEvent } from '@testing-library/react';

// Types
import { NotificationType } from '@/types';

// Components
import NotificationSwitch from '@/components/NotificationSwitch';

describe('NotificationSwitch', () => {
  const handleChange = jest.fn();

  it('renders correctly with default value', () => {
    render(
      <NotificationSwitch
        value={NotificationType.InApp}
        onChange={handleChange}
      />,
    );
    expect(screen.getByText('In-app')).toBeInTheDocument();
    expect(screen.getByText('Email')).toBeInTheDocument();
  });

  it('calls onChange with correct value when toggled', () => {
    render(
      <NotificationSwitch
        value={NotificationType.InApp}
        onChange={handleChange}
      />,
    );

    const emailOption = screen.getByText('Email');
    fireEvent.click(emailOption);

    expect(handleChange).toHaveBeenCalledWith(NotificationType.Email);
  });

  it('updates visual state when toggled', () => {
    render(
      <NotificationSwitch
        value={NotificationType.InApp}
        onChange={handleChange}
      />,
    );
    const emailOption = screen.getByText('Email');
    fireEvent.click(emailOption);

    expect(emailOption).toHaveStyle('color: pastelBlue');
  });

  it('should match snapshot', () => {
    const { container } = render(
      <NotificationSwitch
        value={NotificationType.InApp}
        onChange={handleChange}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
