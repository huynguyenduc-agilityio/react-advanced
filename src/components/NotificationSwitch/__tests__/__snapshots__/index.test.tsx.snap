// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationSwitch should match snapshot 1`] = `
<div>
  <div
    class="css-1oco7to"
  >
    <div
      class="css-1ai9la5"
    >
      <svg
        fill="none"
        height="15"
        viewBox="0 0 14 15"
        width="14"
      >
        <path
          d="M10.0002 0.630859H4.00024C3.60293 0.632505 3.22235 0.791068 2.9414 1.07202C2.66045 1.35297 2.50189 1.73354 2.50024 2.13086V13.1309C2.50189 13.5282 2.66045 13.9088 2.9414 14.1897C3.22235 14.4707 3.60293 14.6292 4.00024 14.6309H10.0002C10.3976 14.6292 10.7781 14.4707 11.0591 14.1897C11.34 13.9088 11.4986 13.5282 11.5002 13.1309V2.13086C11.4986 1.73354 11.34 1.35297 11.0591 1.07202C10.7781 0.791068 10.3976 0.632505 10.0002 0.630859ZM7.00024 4.13086C6.85191 4.13086 6.7069 4.08687 6.58357 4.00446C6.46023 3.92205 6.3641 3.80492 6.30733 3.66787C6.25057 3.53083 6.23572 3.38003 6.26465 3.23454C6.29359 3.08906 6.36502 2.95542 6.46991 2.85053C6.5748 2.74564 6.70844 2.67421 6.85393 2.64527C6.99941 2.61633 7.15021 2.63118 7.28726 2.68795C7.4243 2.74472 7.54143 2.84084 7.62385 2.96418C7.70626 3.08752 7.75024 3.23252 7.75024 3.38086C7.75024 3.57977 7.67123 3.77054 7.53057 3.91119C7.38992 4.05184 7.19916 4.13086 7.00024 4.13086Z"
          fill="currentColor"
        />
      </svg>
      <p
        class="chakra-text css-g5hezf"
      >
        In-app
      </p>
    </div>
    <label
      class="chakra-switch css-16pgy8f"
      hidden=""
    >
      <input
        class="chakra-switch__input"
        style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
        type="checkbox"
        value=""
      />
      <span
        aria-hidden="true"
        class="chakra-switch__track css-14qxnv8"
      >
        <span
          class="chakra-switch__thumb css-0"
        />
      </span>
    </label>
    <div
      class="css-1jg5kzv"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-13otjrl"
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.003 5.884L10 9.882l7.997-3.998A2 2 0 0016 4H4a2 2 0 00-1.997 1.884z"
        />
        <path
          d="M18 8.118l-8 4-8-4V14a2 2 0 002 2h12a2 2 0 002-2V8.118z"
        />
      </svg>
      <p
        class="chakra-text css-g5hezf"
      >
        Email
      </p>
    </div>
  </div>
</div>
`;
